src/bin/pg_autoctl/keeper.c

```diff
@@ -1333,3 +1318,211 @@ keeper_state_as_json(Keeper *keeper, char *json, int size)
+			 "to update HBA rules, including %d changes.",
+			 otherNodesArray->count, diffNodesArray.count);
+
+	/*
+	 * We have a new list of other nodes, update the HBA file. We only update
+	 * the nodes that we didn't know before, or that have a new host property.
+	 */
+	if (!keeper_update_group_hba(keeper, &diffNodesArray))
+	{
+		log_error("Failed to update the HBA entries for the new "
+				  "elements in the our formation \"%s\" and group %d",
+				  keeper->config.formation,
+				  keeper->state.current_group);
+
+		/*
+		 * If we failed to install the changes, retain the previous version of
+		 * the otherNodesArray in-memory (cancel the update).
+		 *
+		 * The otherNodesArrays is also used in the keeper maintenance
+		 * functions for the replication slots, but here we return false and
+		 * then the main keeper loop will skip slot maintenance (part of the
+		 * keeper_ensure_current_state duties).
+		 */
+		*otherNodesArray = oldNodesArray;
+
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * diff_nodesArray computes the array of nodes entries that should be added in
+ * the HBA file in the given pre-allocated diffNodesArray parameter. The diff
+ * is computed from the keeper's otherNodesArray on the previous round, and the
+ * one we just got from the monitor.
+ */
+static void
+diff_nodesArray(NodeAddressArray *previousNodesArray,
+				NodeAddressArray *currentNodesArray,
+				NodeAddressArray *diffNodesArray)
+{
+	int prevIndex = 0;
+	int currIndex = 0;
+	int diffIndex = 0;
+
+	NodeAddress *prevNode = NULL;
+
+	if (previousNodesArray->count == 0)
+	{
+		/* all the entries are new and we want them in diffNodesArray */
+		*diffNodesArray = *currentNodesArray;
+		return;
+	}
+
+	prevNode = &(previousNodesArray->nodes[prevIndex]);
+
+	/* we only care about the nodes in the current nodes array */
+	for (currIndex = 0; currIndex < currentNodesArray->count; currIndex++)
+	{
+		NodeAddress *currNode = &(currentNodesArray->nodes[currIndex]);
+
+		/* remember, the input arrays are sorted on nodeId */
+		if (currNode->nodeId < prevNode->nodeId)
+		{
+			diffNodesArray[diffIndex++] = currentNodesArray[currIndex];
+		}
+		else if (currNode->nodeId == prevNode->nodeId)
+		{
+			if (!streq(currNode->host, prevNode->host))
+			{
+				log_debug("Node %d has a new hostname \"%s\"",
+						  currNode->nodeId, currNode->host);
+
+				diffNodesArray[diffIndex++] = currentNodesArray[currIndex];
+			}
+			prevIndex++;
+		}
+		else if (currNode->nodeId > prevNode->nodeId)
+		{
+			/*
+			 * All the remaining entries of currentNodesArray are new.
+			 *
+			 * We might have entries in previousNodesArray that are not found
+			 * in currentNodesArray anymore, but we don't know how to clean-up
+			 * the HBA file entries at the moment anyway, so we just skip them.
+			 */
+			diffNodesArray[diffIndex++] = currentNodesArray[currIndex];
+			break;
+		}
+		else
+		{
+			log_error("BUG in diff_nodesArray!");
```

Comment:
Minor suggestion:
given that these type of functions are always prone to bugs, one suggestion could be to follow Postgres `list_difference`. And, an example of specific implementation in Citus code can be found [here](https://github.com/citusdata/citus/blob/004e0e4617081205dc48f7f6c58a3a7d4b954f20/src/backend/distributed/planner/multi_physical_planner.c#L5200)  

The  above  suggestion is certainly not  as optimized as this code, but looks  simpler/safer. Given that we don't expect 100s of nodes, this seems like a useful trade-off for me

Reply:
I think I see what you mean, but the simplification you're proposing would still be O(n²), right? I find the current code simple enough at the moment to avoid getting there...